src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -0,0 +1,278 @@
+          [[builtin(workgroup_id)]] group_id : vec3<u32>,
+          [[builtin(num_workgroups)]] num_groups : vec3<u32>,
+        `
+        local_id = 'local_id'
+        local_index = 'local_index'
+        global_id = 'global_id'
+        group_id = 'group_id'
+        num_groups = 'num_groups'
+        break;
+      case 'struct':
+        structures = `struct Inputs {
+            [[builtin(local_invocation_id)]] local_id : vec3<u32>;
+            [[builtin(local_invocation_index)]] local_index : u32;
+            [[builtin(global_invocation_id)]] global_id : vec3<u32>;
+            [[builtin(workgroup_id)]] group_id : vec3<u32>;
+            [[builtin(num_workgroups)]] num_groups : vec3<u32>;
+          };`
+        params = `inputs : Inputs`
+        local_id = 'inputs.local_id'
+        local_index = 'inputs.local_index'
+        global_id = 'inputs.global_id'
+        group_id = 'inputs.group_id'
+        num_groups = 'inputs.num_groups'
+        break;
+      case 'mixed':
+        structures = `struct InputsA {
+          [[builtin(local_invocation_index)]] local_index : u32;
+          [[builtin(global_invocation_id)]] global_id : vec3<u32>;
+        };
+        struct InputsB {
+          [[builtin(workgroup_id)]] group_id : vec3<u32>;
+        };`
+        params = `[[builtin(local_invocation_id)]] local_id : vec3<u32>,
+                  inputsA : InputsA,
+                  inputsB : InputsB,
+                  [[builtin(num_workgroups)]] num_groups : vec3<u32>,`
+        local_id = 'local_id'
+        local_index = 'inputsA.local_index'
+        global_id = 'inputsA.global_id'
+        group_id = 'inputsB.group_id'
+        num_groups = 'num_groups'
+        break;
+    }
+
+    // WGSL shader that stores every builtin value to a buffer, for every invocation in the grid.
+    const wgsl = `
+[[block]]
+struct S {
+  data : array<u32>;
+};
+[[block]]
+struct V {
+  data : array<vec3<u32>>;
+};
+[[group(0), binding(0)]] var<storage, write> local_id_out : V;
+[[group(0), binding(1)]] var<storage, write> local_index_out : S;
+[[group(0), binding(2)]] var<storage, write> global_id_out : V;
+[[group(0), binding(3)]] var<storage, write> group_id_out : V;
+[[group(0), binding(4)]] var<storage, write> num_groups_out : V;
+
+${structures}
+
+let group_width = ${t.params.groupSize.x}u;
+let group_height = ${t.params.groupSize.y}u;
+let group_depth = ${t.params.groupSize.z}u;
+
+[[stage(compute), workgroup_size(group_width, group_height, group_depth)]]
+fn main(
+  ${params}
+  ) {
+  let group_index = ((${group_id}.z * ${num_groups}.y) + ${group_id}.y) * ${num_groups}.x + ${group_id}.x;
+  let global_index = group_index * ${invocationsPerGroup}u + ${local_index};
+  local_id_out.data[global_index] = ${local_id};
+  local_index_out.data[global_index] = ${local_index};
+  global_id_out.data[global_index] = ${global_id};
+  group_id_out.data[global_index] = ${group_id};
+  num_groups_out.data[global_index] = ${num_groups};
+}
+`;
```

Comment:
nit: We usually indent wgsl strings to match the surrounding code for readability (it'll end up with unnecessary spaces inside the string but it doesn't really matter)

Reply:
Done.